bin/build.js

```diff
@@ -57,6 +67,237 @@ async function build() {
         console.log('output length:', srcLength);
         console.log('compression ratio:', 1-(minLength/srcLength));
     }
+
+}
+
+async function hasDirectory(dir, subdir) {
+    return (await fsp.readdir(dir)).includes(subdir) && (await fsp.lstat(path.join(dir, subdir))).isDirectory();
+}
+
+function isObject(type) {
+    return type && type.name && type.name.toLowerCase() === 'object';
+}
+function getTypeString(type) {
+    if (type.name === undefined || type.params === undefined) return type.toString();
+    if (isObject(type)) return 'Object';
+    return type.params.length ? `${type.name}<${type.params.map(getTypeString).join(', ')}>` : type.name;
+}
+function getParamString(param) {
+    const str = param.type ? `${param.name}: ${getTypeString(param.type)}` : param.name;
+    return param.optional ? `${str}?` : str;
+}
+
+const SERVICE_FILTERS = {
+    all: () => true,
+    fsonly: service => service.servicePath,
+};
+function getServiceFilter(filterString = 'fsonly') {
+    const filters = filterString.split(',').map(s => s.trim()).filter(s => s.length).map(s => SERVICE_FILTERS[s]);
+    return s => filters.every(f => f(s));
+}
+
+const SERVICE_DIR_REGEX = /(.*)\/.*\.js/;
+const DOCS_PATH = path.join(__dirname, '..', 'src', 'server', 'docs');
+const GENERATED_PATH = path.join(DOCS_PATH, '_generated');
+
+function getCategories(obj) {
+    const cats = obj.categories;
+    return cats && cats.length ? cats : ['index'];
+}
+function updateCategories(categories, name, obj) {
+    for (const category of getCategories(obj)) {
+        const cat = categories[category];
+        if (cat) cat.items.push(name);
+        else categories[category] = { description: undefined, items: [name] };
+    }
+}
+function sortCategories(categories) {
+    for (const category in categories) {
+        categories[category].items.sort();
+    }
+}
+
+function trimText(str) {
+    str = (str || '').trim();
```

Comment:
If you provide a default value for `str`, you can probably simplify this to just `str.trim()`.

Reply:
Yeah, I tried that initially but apparently sometimes it's null